File: src/main/java/com/vjiki/music/mapper/StoryMapper.java
package com.vjiki.music.mapper;

import java.time.OffsetDateTime;

import com.vjiki.music.dto.StoryResponse;
import com.vjiki.music.entity.Story;

public final class StoryMapper {

    private StoryMapper() {
    }

    public static StoryResponse toResponse(Story story) {
        OffsetDateTime now = OffsetDateTime.now();
        boolean isExpired = story.getExpiresAt() != null && story.getExpiresAt().isBefore(now);

        return StoryResponse.builder()
                .id(story.getId())
                .userId(story.getUserId())
                .userNickname(story.getUser() == null ? null : story.getUser().getNickname())
                .userAvatarUrl(story.getUser() == null ? null : story.getUser().getAvatarUrl())
                .imageUrl(story.getImageUrl())
                .previewUrl(story.getPreviewUrl())
                .storyType(story.getStoryType() == null ? null : story.getStoryType().name())
                .songId(story.getSongId())
                .songTitle(story.getSong() == null ? null : story.getSong().getTitle())
                .songArtist(story.getSong() == null ? null : SongMapper.extractArtist(story.getSong().getArtists()))
                .caption(story.getCaption())
                .location(story.getLocation())
                .viewsCount(story.getViewsCount())
                .createdAt(story.getCreatedAt())
                .expiresAt(story.getExpiresAt())
                .isExpired(isExpired)
                .build();
    }
}

File: bin/main/com/vjiki/music/mapper/StoryMapper.kt
package com.vjiki.music.mapper

import com.vjiki.music.dto.StoryResponse
import com.vjiki.music.entity.Story
import org.springframework.stereotype.Component
import java.time.OffsetDateTime

@Component
object StoryMapper {
    fun Story.toResponse(): StoryResponse {
        val now = OffsetDateTime.now()
        val isExpired = expiresAt?.isBefore(now) ?: false

        return StoryResponse(
            id = id,
            userId = userId,
            userNickname = user?.nickname,
            userAvatarUrl = user?.avatarUrl,
            imageUrl = imageUrl,
            previewUrl = previewUrl,
            storyType = storyType.name,
            songId = songId,
            songTitle = song?.title,
            songArtist = extractArtist(song?.artists),
            caption = caption,
            location = location,
            viewsCount = viewsCount,
            createdAt = createdAt,
            expiresAt = expiresAt,
            isExpired = isExpired
        )
    }

    private fun extractArtist(artists: Map<String, List<String>>?): String? {
        return artists?.get("default")?.firstOrNull()
    }
}

